protocol/storage/miner.go

```diff
@@ -772,102 +772,59 @@ func (sm *Miner) getProvingPeriodStart() (*types.BlockHeight, error) {
-		log.Errorf("PoSt generation time took negative block time: %s < %s", height, start)
-		return
-	}
-
-	if height.GreaterEqual(end) {
-		// TODO: we are too late, figure out faults and decide if we want to still submit
-		log.Errorf("PoSt generation was too slow height=%s end=%s", height, end)
+		log.Errorf("failed to calculate PoSt: %s", err)
 		return
 	}
 
-	// TODO: figure out a more sensible timeout
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
-	defer cancel()
-
-	// TODO: algorithmically determine appropriate values for these
 	gasPrice := types.NewGasPrice(submitPostGasPrice)
-	gasLimit := types.NewGasUnits(submitPostGasLimit)
-
-	_, err = sm.porcelainAPI.MessageSend(ctx, sm.minerOwnerAddr, sm.minerAddr, types.ZeroAttoFIL, gasPrice, gasLimit, "submitPoSt", proofs)
+	_, err = sm.porcelainAPI.MessageSend(ctx, sm.minerOwnerAddr, sm.minerAddr, submission.Fee, gasPrice, submission.GasLimit, "submitPoSt", submission.Proofs)
 	if err != nil {
 		log.Errorf("failed to submit PoSt: %s", err)
 		return
 	}
 
-	log.Debug("submitted PoSt")
+	log.Info("submitted PoSt")
 }
 
-// Query responds to a query for the proposal referenced by the given cid
-func (sm *Miner) Query(ctx context.Context, c cid.Cid) *storagedeal.Response {
-	storageDeal, err := sm.porcelainAPI.DealGet(ctx, c)
+//
+// Implementation of ProofReader.
+//
+
+// ChainHeight provides the chain height to the Prover.
+func (sm *Miner) ChainHeight() (*types.BlockHeight, error) {
+	return sm.porcelainAPI.ChainBlockHeight()
+}
+
+// ChallengeSeed provides a challenge seed to the Prover.
+func (sm *Miner) ChallengeSeed(ctx context.Context, periodStart *types.BlockHeight) (types.PoStChallengeSeed, error) {
+	bytes, err := sm.porcelainAPI.ChainSampleRandomness(ctx, periodStart)
 	if err != nil {
-		return &storagedeal.Response{
-			State:   storagedeal.Unknown,
-			Message: "no such deal",
-		}
+		return types.PoStChallengeSeed{}, errors.Wrap(err, "error sampling chain for randomness")
 	}
 
-	return storageDeal.Response
+	seed := types.PoStChallengeSeed{}
+	copy(seed[:], bytes)
+	return seed, nil
 }
 
-func (sm *Miner) handleQueryDeal(s inet.Stream) {
-	defer s.Close() // nolint: errcheck
-
-	ctx := context.Background()
+//
+// Implementation of ProofCalculator.
+//
 
-	var q storagedeal.QueryRequest
-	if err := cbu.NewMsgReader(s).ReadMsg(&q); err != nil {
-		log.Errorf("received invalid query: %s", err)
-		return
+// CalculatePost calls the sector builder to compute a proof.
+func (sm *Miner) CalculatePost(sortedCommRs proofs.SortedCommRs, seed types.PoStChallengeSeed) ([]types.PoStProof, []uint64, error) {
+	req := sectorbuilder.GeneratePoStRequest{
+		SortedCommRs:  sortedCommRs,
+		ChallengeSeed: seed,
 	}
-
-	resp := sm.Query(ctx, q.Cid)
-
-	if err := cbu.NewMsgWriter(s).WriteMsg(resp); err != nil {
-		log.Errorf("failed to write query response: %s", err)
+	res, err := sm.node.SectorBuilder().GeneratePoSt(req)
```

Comment:
Miner only needs SectorBuilder and BlockService from node.  I wonder if there is a good way to do this where miner doesn't have to access node.  I'm saying this b/c I'm not sure how much I like having Miner be the default ProofCalculator and ProofReader rather than their being separate (either separate structs or in a separate package), since the SectorBuilder is the one that generates the PoSt. It feels a little like mixing concerns, OTOH what else is concerned with coming up with proofs but a Miner....